### PR TITLE
chore(meta): Update pull request guidelines

### PR DESCRIPTION
**Pull request by bsmth**
This PR reworks the PR guidelines doc with some additional sections:

- Before you start
- Values and participation
- Reviewers and assignees

TBD:

- Topic and review owners / teams / alumni section

### CODE REVIEW

**Comment by wbamberg** [comment]
PRs are auto-assigned to reviewers based on CODEOWNERS, which itself uses teams and (I guess randomly) assigns a PR to a member of a team. If I see a PR that's assigned to someone else as a reviewer, can I adopt it? That is, can I review it and merge it without the involvement of the originally assigned reviewer?

Because this is a pretty common practice right now. It seems like I can't do this any more? I can review the PR if I feel like it, but not merge it, because:

>  After submitting your review, you can safely merge the pull request **if there are no other reviewers** or any outstanding review comments to address.

If this is the case I'm pretty likely to stop reviewing PRs that I'm not assigned as a reviewer on, because I can't resolve them. And I think that as a rule will make us a lot less efficient.

At the moment we have this kind of chaotic "just pick a PR" approach. I think there are a couple of possible problems with it, both of which can have lighter solutions than this process:

1. One is that we might feel that some PRs are specialised and not everyone is qualified to review them. I think this is sometimes true but it is dependent not just on the area of the site but on the specific PR - for example a typo fix, or some language work, might be reviewable by anyone even in a very niche area of the site. I think historically we do this case-by-case by using our judgement, and it works pretty well - people are sensible enough not to review things they don't understand well.

Even if this wasn't the case though, reviewers are (AIUI) **randomly** assigned to members of a team. Since it was random in the first place, what is wrong with someone else in the team being the person to review and merge the PR? It might have been assigned to them anyway, on a different day. As I understand it this process disallows that, but I'm not sure why.

2. Two is that, especially for long/involved PRs, we really want to avoid race conditions where two people review something at the same time, and the winner merges the PR, and the loser loses all the time they spent reviewing it. This is especially bad because good reviews take time, so will usually lose.

This doc already mentions one way to avoid this problem, by making the reviewer comment before they are going to start, and that would be OK. An alternative is that we **don't** auto-assign reviewers. Either the author manually assigns a reviewer (which covers the case where a PR needs to be reviewed by a specific person) or, when someone picks up a review, they assign themselves, which avoids races.

File: files/en-us/mdn/community/pull_requests/index.md [hunk 1]
```diff
@@ -10,127 +10,194 @@ tags:
...
+
+If you are still not sure or if you think the content is suspicious, reach out to the MDN Web Docs team and ask for help.
+
+## Reading list
+
+Reviewers are encouraged to read the following articles for help with common reviewer tasks:
+
+- [The Art of Closing: How to closing an unfinished or rejected pull request](https://blog.jessfraz.com/post/the-art-of-closing/)
+- [Kindness and Code Reviews: Improving the Way We Give Feedback](https://product.voxmedia.com/2018/8/21/17549400/kindness-and-code-reviews-improving-the-way-we-give-feedback)
+- [Code Review Guidelines for Humans: Examples of good and back feedback](https://phauer.com/2018/code-review-guidelines/#code-reviews-guidelines-for-the-reviewer)
```
Comment (dipikabh) [inline files/en-us/mdn/community/pull_requests/index.md:176]:
```suggestion
- [The Art of Closing](https://blog.jessfraz.com/post/the-art-of-closing/) explains how to close an unfinished or rejected pull request
- [Kindness and Code Reviews: Improving the Way We Give Feedback](https://product.voxmedia.com/2018/8/21/17549400/kindness-and-code-reviews-improving-the-way-we-give-feedback) gives useful hints to give feedback
- [Code Review Guidelines for the Reviewer](https://phauer.com/2018/code-review-guidelines/#code-reviews-guidelines-for-the-reviewer) provides examples of good and bad feedback
```

File: files/en-us/mdn/community/pull_requests/index.md [hunk 2]
```diff
@@ -10,127 +10,194 @@ tags:
...
+
+## Timelines
+
+This section provides details for expected turnaround times while reviewing pull requests if you're a reviewer and while responding to review comments if you're a pull request author.
```
Comment (dipikabh) [inline files/en-us/mdn/community/pull_requests/index.md:180]:
```suggestion
### Guidelines for turnaround times for authors and reviewers

This section provides details for expected turnaround times while responding to review comments if you're a pull request author and while reviewing pull requests if you're a reviewer.
```

File: files/en-us/mdn/community/pull_requests/index.md [hunk 3]
```diff
@@ -10,127 +10,194 @@ tags:
...
+
+- **Reviews**:
```
Comment (dipikabh) [inline files/en-us/mdn/community/pull_requests/index.md:182]:
```suggestion
- **Reviewing**:
```

File: files/en-us/mdn/community/pull_requests/index.md [hunk 4]
```diff
@@ -10,127 +10,194 @@ tags:
...
+  If you must create a new PR, it can reference the closed one.
 
-## Topic review owners
+## Pull request review process
 
-The following specific topic areas are being reviewed by the kind souls listed underneath them. Be kind to them, and thank them for all the help they give to this project. If you would like to help with MDN content reviews,
-[get in touch with us](/en-US/docs/MDN/Community/Contributing/Getting_started#Step_4_Ask_for_help).
+We add reviewers automatically based on a `CODEOWNERS` file but, if there is a specific person you want to request review from, manually [request a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review).
+We also use auto-labeling on pull requests to help triage.
+Maintainers will triage pull requests and add any additional labels, such as `needs-info` or `on-hold`, if needed based on context.
```
Comment (dipikabh) [inline files/en-us/mdn/community/pull_requests/index.md:97]:
```suggestion
Reviewer(s) are automatically assigned when you open a pull request based on a `CODEOWNERS` file, but if there is a specific person you want to request review from, you can [request a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) manually.
We also use auto-labeling on pull requests to help us triage them.
Maintainers can further triage pull requests and add any additional labels, such as `needs-info` or `on-hold`, if needed based on context.
```

**Review by dipikabh — CHANGES_REQUESTED** [review]
Thanks @bsmth for porting the changes here! Like I said before, the more times I look at this page, I feel it can be improved a bit more. 😄 

Looking forward to see what @wbamberg thinks about this page.

**Comment by dipikabh** [comment]
@wbamberg, I believe both of your comments can be handled by using the "Assignees" field and by commenting to let the assigned reviewer know that you are starting/taking on the PR review. If you add yourself to "Assignees", you can assume the responsibility of approving and merging.

I have had same-time-review conflict even on short PRs. We could avoid the review conflict situation by making it a practice to leave a comment before starting a review, even if we are the assigned reviewer. So that somebody else who comes across the PR would know that the assigned reviewer has started the review.

File: files/en-us/mdn/community/pull_requests/index.md [hunk 5]
```diff
@@ -10,127 +10,194 @@ tags:
...
-  - For more complex merge conflicts, you should use the [command line(terminal)](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/addressing-merge-conflicts/resolving-a-merge-conflict-using-the-command-line) to resolve the conflicts.
-- **Don't reopen closed PRs**: Don't re-open a PR that a reviewer has closed unless there has been a discussion and a consensus reached to do so. In most cases it is best to open a new PR and reference the previous PR.
+- **Day-to-day improvements:** for the documentation of APIs, CSS properties, platform updates and content additions.
+  This is usually done by MDN Web Docs staff working for Mozilla, Google, Open Web Docs, Samsung, but also by community volunteers.
+- **Minor fixes:** and small updates to the site for fixing typos, grammatical issues, and technical inaccuracies.
+  These issues are usually found by readers of MDN Web Docs.
+- **Content bug fixes:** usually done by volunteers to close issues on the `mdn/content` repo.
```
[the remark this answers is not in the document]
Comment (bsmth) [reply files/en-us/mdn/community/pull_requests/index.md:20]: Good, I prefer this 👍🏻

File: files/en-us/mdn/community/pull_requests/index.md [hunk 6]
```diff
@@ -10,127 +10,194 @@ tags:
...
+A good place to start is by asking for information from the PR author to help.
+
+It is rare that you'll be required to review a large, complex content change with no warning.
+If this does happen, however, the PR description should link to an issue that explains the background information.
+
+If you are still not sure or if you think the content is suspicious, reach out to the MDN Web Docs team and ask for help.
+
+## Reading list
+
+Reviewers are encouraged to read the following articles for help with common reviewer tasks:
```
[the remark this answers is not in the document]
Comment (bsmth) [reply files/en-us/mdn/community/pull_requests/index.md:172]: Yep, let's do this

File: files/en-us/mdn/community/pull_requests/index.md [hunk 7]
```diff
@@ -10,127 +10,194 @@ tags:
...
+## Values and participation
 
-Note: You may encounter merge conflicts as you review pull requests, if an another pull request that touches some of the same files got merged before the one you are reviewing. [Addressing merge conflicts](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/addressing-merge-conflicts) is a useful resource to help you. Feel free also to ask your team(s) for help if you need it.
+We want MDN Web Docs to be a welcoming, friendly community that we can all be proud of.
+All participants must follow our [Code of Conduct](https://github.com/mdn/content/blob/main/CODE_OF_CONDUCT.md) which means adhering to [Mozilla's Community Participation Guidelines](https://www.mozilla.org/en-US/about/governance/policies/participation/).
+Be polite and constructive when opening pull requests, writing review comments, interacting with the PR author or other community members.
+If anyone has engaged in behavior that is potentially illegal or makes you or someone else feel unsafe, unwelcome, or uncomfortable, you are encouraged to [report it](https://www.mozilla.org/en-US/about/governance/policies/participation/reporting/).
```
[the remark this answers is not in the document]
Comment (bsmth) [reply files/en-us/mdn/community/pull_requests/index.md]: Done in 6d188b26cdf4dde9bae69e4b845a4ec02147868f

File: files/en-us/mdn/community/pull_requests/index.md [hunk 8]
```diff
@@ -10,127 +10,194 @@ tags:
...
+- **Reviews**:
+  The pull request reviewer should be able to review the changes in 2 weeks or less.
+  In the 2 weeks after a pull request is open, the reviewer can:
+  - Leave a comment about when they can start reviewing the pull request
+  - Ask for technical or resource help
+- **Addressing requested changes:**
+  The PR author should be able to respond to or fix the comments in 4 weeks or less.
+  If the PR author is unable to respond or fix the review comments in that time, the reviewer can do one of the following:
+  - Commit the changes and merge the PR
+  - Close the PR
```
[the remark this answers is not in the document]
Comment (bsmth) [reply files/en-us/mdn/community/pull_requests/index.md]: Done in 6d188b26cdf4dde9bae69e4b845a4ec02147868f

**Comment by bsmth** [comment]
Thanks for the feedback, Will. To your question:

>If I see a PR that's assigned to someone else as a reviewer, can I adopt it? That is, can I review it and merge it without the involvement of the originally assigned reviewer?

You can, indeed. In the [PR process section](https://github.com/mdn/content/blob/6d188b26cdf4dde9bae69e4b845a4ec02147868f/files/en-us/mdn/community/pull_requests/index.md#pull-request-review-process)

> If you want to review a pull request but are not listed as a reviewer, you may add yourself as one. It's polite to check with existing reviewers first by commenting on the pull request that you intend to start a review.

**Comment by wbamberg** [comment]
> @wbamberg, I believe both of your comments can be handled by using the "Assignees" field and by commenting to let the assigned reviewer know that you are starting/taking on the PR review. If you add yourself to "Assignees", you can assume the responsibility of approving and merging.

For example. https://github.com/mdn/content/pull/22905 currently has no assignees, and has four reviewer assigned, none of them me:

```
@mdn/yari-content-css
@mdn/yari-content-mdn
@sideshowbarker
@estelle
```

(or maybe `@mdn/yari-content-mdn`  is implicitly me?)

If I want to "adopt" this PR - by which I mean, review it and merge it after I am happy, without waiting for any other reviews - can I, and if so how?

Maybe:
- add a comment saying that I'm going to review it
- review it
- once I'm happy, merge it
?

But as above, this seems to fall foul of:

> After submitting your review, you can safely merge the pull request **if there are no other reviewers** or any outstanding review comments to address.

...because there are outstanding reviewers. So should I also remove existing reviewers, when I add myself? Or should the quoted rule be revised?

I'm generally a bit puzzled about the workflow wrt assignees. It seems that we expect PRs to not have assignees initially, right? But the only place the process talks about setting an assignee is "You may set the author as an assignee to indicate that a pull request needs their attention before a review can start." There's no indication of when or why we might set the assignee to anyone else.

It's also odd to me for "assignee" to have these (at least) 2 functions of "person who does the work" (which is generally the author) and "person who has the right to merge the PR" (which is usually not the author).

Sorry if this all sounds pedantic.

> I have had same-time-review conflict even on short PRs. We could avoid the review conflict situation by making it a practice to leave a comment before starting a review, even if we are the assigned reviewer. So that somebody else who comes across the PR would know that the assigned reviewer has started the review.

Agreed that this also happens, I don't care in that case though because I didn't lose a lot of work. Commenting is also fine as a signal. I think adding self as a reviewer is a better signal because it is more formal than a comment. We can define "adding yourself as a reviewer" to mean exactly "don't merge this PR without my review", whereas a comment is fuzzier, and also easier to miss, given that we also have auto-added comments like PR preview links. However if reviewers are auto-assigned we can't use reviewer in this way :(.

**Comment by dipikabh** [comment]
> If I want to "adopt" this PR - by which I mean, review it and merge it after I am happy, without waiting for any other reviews - can I, and if so how?
> 
> Maybe:
> 
>     add a comment saying that I'm going to review it
>     review it
>     once I'm happy, merge it
>     ?

Yes, you should be able to do that by adding yourself to "Assignees" (this lets other assigned reviewers know that you're taking the lead on the PR) and leaving a comment that you're starting the review. If within the time of your review and author's fixes, nobody else has left a comment that needs to be fixed, you can feel free to approve and merge. We have been doing this anyway, only now we are trying to a more controlled system in place.

Like you're suggesting, we should update the text to:

> After submitting your review, you can safely merge the pull request if there are no outstanding review comments to be addressed.

In cases where you are the only reviewer, it would be optional to add yourself to "Assignees", but it would be nice to comment if you've started the review (I have been forgetting to do this myself) so that somebody who is not on the reviewer list but is interested in reviewing the PR knows that the PR is being actively looked at. Anybody can still leave a comment/review if you've already started a review but they would at least be aware that there could be overlaps in feedback.

Yeah, reviewers are automatically assigned and sometimes when there multiple reviewers and long discussions, nobody has the ownership to take the PR to closure. For all these reasons and to avoid review conflicts, we've been discussing using the "Assignees" field more.

> It seems that we expect PRs to not have assignees initially, right? But the only place the process talks about setting an assignee is in the bit on CI failures and merge conflicts "You may set the author as an assignee to indicate that a pull request needs their attention before a review can start." There's no indication of when or why we might set the assignee to anyone else.

That's correct. And if the situation requires (in case of multiple reviewers), we would need to add ourselves to "Assignees". I personally feel that we should restrict the use of "Assignees" to us reviewers so as not to over complicate the workflow. If a pull request needs an author's attention before a review can start, we should let the author know through a comment. 

The purpose of using "Assignees" is to help us when there are multiple reviewers and to avoid review conflicts.

**Comment by bsmth** [comment]
> > After submitting your review, you can safely merge the pull request **if there are no other reviewers** or any outstanding review comments to address.
> 
> ...implies that I **can't** merge it after my review (because there will still be the original review request).

Sorry I misunderstood the point, thanks for elaborating on it. This is moreso "criteria for merging" -> we anyhow have repo configuration that at least one approving review is a hard requirement. "At least 1 approving review is required by reviewers with write access." the rest should be left to our judgement (including dismissing a stale review, approving and merging).

Re: assignees - GitHub doesn't define what an assignee is, so it's up to maintainers to decide how this makes sense. For us, I can see that "assignee" is useful to know at-a-glance who is responsible for next steps on a PR, this is especially useful for old / stalled PRs of which we have quite a few - it takes time to read through each to see "what's the context? what are next steps? who will move it forward? who does it depend on?" - this can be outcome/decision from external projects, teams, or systems. 

A good example is assigning Schalk who is usually neither the reviewer nor the author. Assigning Schalk in this case signals that it needs feedback or action from his side before we can proceed with anything else. Adding someone as an assignee can also be used to indicate that they take responsibility for a stale PR, finishing work started by the original author.

We can certainly riff on variations of this:

> you can safely merge the pull request if there are no outstanding changes requested.

With Dipika's point:

>I personally feel that we should restrict the use of "Assignees" to us reviewers so as not to over complicate the workflow. 

If it's a stale PR that the author needs to address changes, assigning them makes sense, but I see where you're coming from.

> The purpose of using "Assignees" is to help us when there are multiple reviewers and to avoid review conflicts.

When there are multiple reviewers, we could either comment "Going to review this now" or assign it to yourself, indicating that you want to take ownership, that's up for decision.

**Pull request closed without merging**